[FLINK-32204][runtime] Separates the DirectExecutorService from the DirectExecutor again.

XComp opened the pull request:
## What is the purpose of the change

The `DirectExecutor` doesn't have to handle the shutdown state and, therefore, can be served as a singleton. The `DirectExecutorService`, instead, cannot be served as a singleton due to the shutdown state.

We still have to provide an `ExecutorService` implementation without proper task rejection after shutdown was called, though, due to a bug in the `TreeCache` implementation as described in FLINK-32204.

## Brief change log

* Merges TreeCache instantiation of `ZooKeeperUtils` and `ZooKeeperMultipleComponentLeaderElectionDriver`
* reintroduced `Executor.DirectExecutor` after it was removed in 17082604
* Introduces new factory methods for the `DirectExecutorService` and made them available through the `Executors` interface

## Verifying this change

* The error couldn't be reproduced locally and is, therefore, hard to verify because it's supposed to be a bug within curator. But the code itself is back to its pre-FLINK-31995 state for the ZooKeeper codebase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

dmvk commented on flink-core/src/main/java/org/apache/flink/util/concurrent/DirectExecutorService.java:
```diff
@@ -39,10 +41,37 @@
+     * Creates a {@code DirectExecutorService} without proper shutdown contract execution, i.e.
+     * shutting down the executor doesn't have an influence on the execution of tasks.
+     */
+    static DirectExecutorService createDirectExecutorServiceWithoutTaskRejection() {
```
"TaskRejection" is an implementation detail; this should be named along the lines of "missing lifecycle" / "always on" ...

dmvk commented on flink-core/src/main/java/org/apache/flink/util/concurrent/DirectExecutorService.java:
```diff
@@ -39,10 +41,37 @@
 class DirectExecutorService implements ExecutorService {
-    static final DirectExecutorService INSTANCE = new DirectExecutorService();
+
+    private final ThrowingRunnable<? extends RuntimeException> ifShutdownAction;
```
Wouldn't a boolean flag do the trick as well + make this easier to interpret? :)

dmvk reviewed (approved):
LGTM 👍 I've left few nits

XComp commented:
Thanks, @dmvk. I addressed your comments and added a `DirectExecutorServiceTest` to cover the behavior. I will merge the PR after another green CI run
